beluga_tutorial/src/main.cpp

```diff
@@ -0,0 +1,333 @@
+  beluga::TupleVector<Particle> reweight;
+  beluga::TupleVector<Particle> resample;
+  Estimate estimation;
+};
+
+struct TutorialDataset {
+  LandmarkMapVector landmark_map;
+  std::vector<TutorialData> sim_data;
+};
+
+void load_params_from_yaml(const std::filesystem::path& path, TutorialParams& tutorial_params) {
+  try {
+    YAML::Node params = YAML::LoadFile(path);
+    tutorial_params.load_landmark_map = params["load_landmark_map"].as<bool>();
+    tutorial_params.save_landmark_map = params["save_landmark_map"].as<bool>();
+    tutorial_params.map_size = params["map_size"].as<int>();
+    tutorial_params.number_of_doors = params["number_of_doors"].as<int>();
+    tutorial_params.number_of_particles = params["number_of_particles"].as<int>();
+    tutorial_params.number_of_cycles = params["number_of_cycles"].as<int>();
+    tutorial_params.initial_pose = params["initial_pose"].as<double>();
+    tutorial_params.initial_pose_sd = params["initial_pose_sd"].as<double>();
+    tutorial_params.dt = params["dt"].as<double>();
+    tutorial_params.velocity = params["velocity"].as<double>();
+    tutorial_params.translation_sd = params["translation_sd"].as<double>();
+    tutorial_params.sensor_range = params["sensor_range"].as<double>();
+    tutorial_params.sensor_model_sigam = params["sensor_model_sigam"].as<double>();
+    tutorial_params.min_particle_weight = params["min_particle_weight"].as<double>();
+    tutorial_params.dataset_path = params["dataset_path"].as<std::string>();
+    tutorial_params.landmark_map_file_name = params["landmark_map_file_name"].as<std::string>();
+  } catch (YAML::BadFile& e) {
+    std::cout << e.what() << "\n";
+  }
+}
+
+void load_landmark_map_from_yaml(const std::filesystem::path& path, LandmarkMapVector& landmark_map) {
+  try {
+    YAML::Node node = YAML::LoadFile(path);
+    landmark_map = node["landmark_map"].as<LandmarkMapVector>();
+  } catch (YAML::BadFile& e) {
+    std::cout << e.what() << "\n";
+  }
+}
+
+void save_landmark_map_to_yaml(const std::filesystem::path& path, LandmarkMapVector& landmark_map) {
+  YAML::Node node;
+  for (const auto& lmp : landmark_map) {
+    node["landmark_map"].push_back(lmp);
+  }
+  std::ofstream fout(path, std::ios::trunc);
+  fout << node;
+  fout.close();
+}
+
+void save_tutorial_dataset_to_yaml(const std::filesystem::path& path, const TutorialDataset& tutorial_dataset) {
+  YAML::Node node;
+
+  for (const auto& lmp : tutorial_dataset.landmark_map) {
+    node["landmark_map"].push_back(lmp);
+  }
+
+  for (auto&& [cycle, data] : tutorial_dataset.sim_data | ranges::views::enumerate) {
+    node["simulation_dataset"][cycle]["ground_truth"] = data.ground_truth;
+
+    for (const auto& current : data.current) {
+      node["simulation_dataset"][cycle]["particles"]["current"]["states"].push_back(std::get<0>(current));
+      node["simulation_dataset"][cycle]["particles"]["current"]["weights"].push_back(
+          static_cast<double>(std::get<1>(current)));
+    }
```

Comment:
```suggestion
    auto current = node["simulation_dataset"][cycle]["particles"]["current"];
    current["states"] = beluga::views::states(data.current) | ranges::to<std::vector<double>>;
    current["weights"]  = beluga::views::weights(data.current) | ranges::to<std::vector<double>>;
```

Reply:
using `auto &current` it throwing me the following error: Non-const lvalue reference to type 'Node' cannot bind to a temporary of type 'Node'. So to use this way I'll need to just copy that using `auto current`. For the purpose of the tutorial is good enough I think

Reply:
Ah, interesting, I thought `[]` returned a reference to a node. I guess that's no the case, it returns a value.

Reply:
Apparently yes, now I'm pushing the modifications.